pkg/operator/boundsatokensignercontroller/controller.go

```diff
@@ -0,0 +1,296 @@
+
+	// queue only ever has one item, but it has nice error handling backoff/retry semantics
+	queue workqueue.RateLimitingInterface
+}
+
+func NewBoundSATokenSignerController(
+	kubeInformersForNamespaces v1helpers.KubeInformersForNamespaces,
+	kubeClient kubernetes.Interface,
+	eventRecorder events.Recorder,
+
+) *BoundSATokenSignerController {
+
+	ret := &BoundSATokenSignerController{
+		secretClient:    v1helpers.CachedSecretGetter(kubeClient.CoreV1(), kubeInformersForNamespaces),
+		configMapClient: v1helpers.CachedConfigMapGetter(kubeClient.CoreV1(), kubeInformersForNamespaces),
+		eventRecorder:   eventRecorder.WithComponentSuffix("bound-sa-token-signer-controller"),
+
+		cachesSynced: []cache.InformerSynced{
+			kubeInformersForNamespaces.InformersFor(operatorNamespace).Core().V1().Secrets().Informer().HasSynced,
+			kubeInformersForNamespaces.InformersFor(targetNamespace).Core().V1().Secrets().Informer().HasSynced,
+			kubeInformersForNamespaces.InformersFor(targetNamespace).Core().V1().ConfigMaps().Informer().HasSynced,
+		},
+
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "BoundSATokenSignerController"),
+	}
+
+	kubeInformersForNamespaces.InformersFor(operatorNamespace).Core().V1().Secrets().Informer().AddEventHandler(ret.eventHandler())
+	kubeInformersForNamespaces.InformersFor(targetNamespace).Core().V1().Secrets().Informer().AddEventHandler(ret.eventHandler())
+	kubeInformersForNamespaces.InformersFor(targetNamespace).Core().V1().ConfigMaps().Informer().AddEventHandler(ret.eventHandler())
+
+	return ret
+}
+
+func (c *BoundSATokenSignerController) sync() error {
+	signingSecret, err := c.secretClient.Secrets(operatorNamespace).Get(NextSigningKeySecretName, metav1.GetOptions{})
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+	needKeypair := errors.IsNotFound(err) || len(signingSecret.Data[PrivateKeyKey]) == 0 || len(signingSecret.Data[PublicKeyKey]) == 0
+	if needKeypair {
+		newSecret, err := newSigningSecret()
+		if err != nil {
+			return err
+		}
+		signingSecret, _, err = resourceapply.ApplySecret(c.secretClient, c.eventRecorder, newSecret)
+		if err != nil {
+			return err
+		}
+		// requeue for after we should have recovered
+		c.queue.AddAfter(workQueueKey, readyInterval+10*time.Second)
+	}
+
+	certConfigMap, err := c.configMapClient.ConfigMaps(targetNamespace).Get(CertConfigMapName, metav1.GetOptions{})
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+	if errors.IsNotFound(err) {
+		certConfigMap = &corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: targetNamespace,
+				Name:      CertConfigMapName,
+			},
+			Data: map[string]string{},
+		}
+	}
+	currPublicKey := string(signingSecret.Data[PublicKeyKey])
+	hasThisKey := false
+	for _, value := range certConfigMap.Data {
+		if value == currPublicKey {
+			hasThisKey = true
+			break
+		}
+	}
+	if !hasThisKey {
+		// Increment until a unique name is found
```

Comment:
this is no pruning?

Reply:
Why would you suggest pruning? And how would you suggest that the operator determine which public keys are safe to delete? As per the example of the equivalent kcmo operator, I don't think pruning is something we can safely do automatically.

Reply:
Updated comment, PTAL